File: ch2/clip10/boj1194/Main.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

class Main {
    static int n, m;
    static char[][] maze;
    static int[][][] distance;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};
    static final int DOOR = 1;
    static final int KEY = 2;
    static final int EXIT = 3;
    static final int WALL = 4;
    static final int NONE = 5;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        maze = new char[n][m];
        distance = new int[n][m][1 << 6];
        Queue<Point> q = new LinkedList<>();
        for (int r = 0; r < n; r++) {
            String line = sc.next();
            for (int c = 0; c < m; c++) {
                char temp = line.charAt(c);
                maze[r][c] = temp;
                if (temp == '0') {
                    q.add(new Point(r, c, 0));
                    distance[r][c][0] = 1;
                }

            }
        }
        while (!q.isEmpty()) {
            Point now = q.poll();
            for (int i = 0; i < 4; i++) {
                int nr = now.r + dr[i];
                int nc = now.c + dc[i];
                if (isOutOfRange(nr, nc)) continue;
                int next = maze[nr][nc];
                switch (getType(next)) {
                    case WALL -> {
                        continue;
                    }
                    case NONE -> {
                        if (distance[nr][nc][now.key] == 0) {
                            distance[nr][nc][now.key] = distance[now.r][now.c][now.key] + 1;
                            q.add(new Point(nr, nc, now.key));
                        }
                    }
                    case EXIT -> {
                        System.out.println(distance[now.r][now.c][now.key]);
                        return;
                    }
                    case KEY -> {
                        int nextKey = now.key | (1 << (next - 'a'));
                        if (distance[nr][nc][nextKey] == 0) {
                            distance[nr][nc][nextKey] =
                                    distance[now.r][now.c][now.key] + 1;
                            q.add(new Point(nr, nc, nextKey));
                        }
                    }
                    case DOOR -> {
                        if ((now.key & (1 << (next - 'A'))) == 0) continue;
                        if (distance[nr][nc][now.key] == 0) {
                            distance[nr][nc][now.key] =
                                    distance[now.r][now.c][now.key] + 1;
                            q.add(new Point(nr, nc, now.key));
                        }
                    }
                }
            }
        }
        System.out.println(-1);
    }

    static boolean isOutOfRange(int r, int c) {
        return r < 0 || r >= n || c < 0 || c >= m;
    }

    static int getType(int c) {
        if (c == '#') return WALL;
        else if (c == '.' || c == '0') return NONE;
        else if (c >= 'a' && c <= 'f') return KEY;
        else if (c >= 'A' && c <= 'F') return DOOR;
        else if (c == '1') return EXIT;
        else return -1;
    }
}

class Point {
    int r, c, key;

    Point(int r, int c, int k) {
        this.r = r;
        this.c = c;
        this.key = k;
    }
}

File: ch2/clip10/boj1194/main.py
import sys
from collections import deque

n, m = map(int, input().split())
maze = [list(input().strip()) for _ in range(n)]
distance = [[[0] * (1 << 6) for _ in range(m)] for _ in range(n)]

dr = [-1, 1, 0, 0]
dc = [0, 0, -1, 1]

def is_out_of_range(r, c):
    return r < 0 or r >= n or c < 0 or c >= m

def get_type(c):
    if c == '#':
        return 4  # WALL
    elif c == '.' or c == '0':
        return 5  # NONE
    elif 'a' <= c <= 'f':
        return 2  # KEY
    elif 'A' <= c <= 'F':
        return 1  # DOOR
    elif c == '1':
        return 3  # EXIT
    else:
        return -1

q = deque()
for r in range(n):
    for c in range(m):
        if maze[r][c] == '0':
            q.append((r, c, 0))
            distance[r][c][0] = 1

while q:
    r, c, key = q.popleft()
    for i in range(4):
        nr, nc = r + dr[i], c + dc[i]
        if is_out_of_range(nr, nc):
            continue
        next_type = get_type(maze[nr][nc])
        if next_type == 4:  # WALL
            continue
        elif next_type == 5:  # NONE
            if distance[nr][nc][key] == 0:
                distance[nr][nc][key] = distance[r][c][key] + 1
                q.append((nr, nc, key))
        elif next_type == 3:  # EXIT
            print(distance[r][c][key])
            sys.exit(0)
        elif next_type == 2:  # KEY
            next_key = key | (1 << (ord(maze[nr][nc]) - ord('a')))
            if distance[nr][nc][next_key] == 0:
                distance[nr][nc][next_key] = distance[r][c][key] + 1
                q.append((nr, nc, next_key))
        elif next_type == 1:  # DOOR
            if (key & (1 << (ord(maze[nr][nc]) - ord('A')))) != 0:
                if distance[nr][nc][key] == 0:
                    distance[nr][nc][key] = distance[r][c][key] + 1
                    q.append((nr, nc, key))

print(-1)
